File: controllers/farmerController.js
const pool = require('../config/db');

// GET /api/farmer/dashboard
exports.getDashboardStats = async (req, res) => {
  try {
    const farmerId = req.user.id;

    const [totalProducts] = await pool.query('SELECT COUNT(*) as count FROM products WHERE farmer_id = ?', [farmerId]);
    const [activeProducts] = await pool.query('SELECT COUNT(*) as count FROM products WHERE farmer_id = ? AND status = "active"', [farmerId]);
    const [totalOrders] = await pool.query(
      'SELECT COUNT(*) as count FROM order_items WHERE farmer_id = ?', [farmerId]
    );
    const [pendingOrders] = await pool.query(
      'SELECT COUNT(*) as count FROM order_items WHERE farmer_id = ? AND status = "pending"', [farmerId]
    );
    const [earnings] = await pool.query(
      `SELECT COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) as total
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       WHERE oi.farmer_id = ? AND o.payment_status = 'paid'`,
      [farmerId]
    );

    // Recent orders
    const [recentOrders] = await pool.query(
      `SELECT oi.*, o.created_at as order_date, o.status as order_status,
        p.name as product_name, u.name as buyer_name
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       JOIN products p ON oi.product_id = p.id
       JOIN users u ON o.buyer_id = u.id
       WHERE oi.farmer_id = ?
       ORDER BY o.created_at DESC LIMIT 5`,
      [farmerId]
    );

    res.json({
      stats: {
        totalProducts: totalProducts[0].count,
        activeProducts: activeProducts[0].count,
        totalOrders: totalOrders[0].count,
        pendingOrders: pendingOrders[0].count,
        totalEarnings: earnings[0].total
      },
      recentOrders
    });
  } catch (err) {
    console.error('Farmer dashboard error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/farmer/inventory
exports.getInventory = async (req, res) => {
  try {
    const farmerId = req.user.id;

    const [products] = await pool.query(
      `SELECT p.*,
        (SELECT pi.image_url FROM product_images pi WHERE pi.product_id = p.id AND pi.is_primary = 1 LIMIT 1) as image
       FROM products p
       WHERE p.farmer_id = ?
       ORDER BY p.quantity ASC`,
      [farmerId]
    );

    const totalItems = products.length;
    const lowStock = products.filter(p => p.quantity > 0 && p.quantity <= 10).length;
    const outOfStock = products.filter(p => p.quantity <= 0).length;

    res.json({
      stats: { totalItems, lowStock, outOfStock },
      products
    });
  } catch (err) {
    console.error('Get inventory error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

// GET /api/farmer/earnings
exports.getEarnings = async (req, res) => {
  try {
    const farmerId = req.user.id;

    // Total earnings
    const [total] = await pool.query(
      `SELECT COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) as total
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       WHERE oi.farmer_id = ? AND o.payment_status = 'paid'`,
      [farmerId]
    );

    // This month
    const [thisMonth] = await pool.query(
      `SELECT COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) as total
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       WHERE oi.farmer_id = ? AND o.payment_status = 'paid'
       AND MONTH(o.created_at) = MONTH(NOW()) AND YEAR(o.created_at) = YEAR(NOW())`,
      [farmerId]
    );

    // Pending
    const [pending] = await pool.query(
      `SELECT COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) as total
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       WHERE oi.farmer_id = ? AND o.payment_status = 'pending'`,
      [farmerId]
    );

    // Recent transactions
    const [transactions] = await pool.query(
      `SELECT oi.*, o.created_at as order_date, o.payment_status,
        p.name as product_name, u.name as buyer_name,
        (oi.price_at_purchase * oi.quantity) as amount
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       JOIN products p ON oi.product_id = p.id
       JOIN users u ON o.buyer_id = u.id
       WHERE oi.farmer_id = ?
       ORDER BY o.created_at DESC LIMIT 10`,
      [farmerId]
    );

    // Monthly earnings (last 6 months)
    const [monthlyEarnings] = await pool.query(
      `SELECT DATE_FORMAT(o.created_at, '%Y-%m') as month,
        COALESCE(SUM(oi.price_at_purchase * oi.quantity), 0) as earnings
       FROM order_items oi
       JOIN orders o ON oi.order_id = o.id
       WHERE oi.farmer_id = ? AND o.payment_status = 'paid'
       AND o.created_at >= DATE_SUB(NOW(), INTERVAL 6 MONTH)
       GROUP BY DATE_FORMAT(o.created_at, '%Y-%m')
       ORDER BY month ASC`,
      [farmerId]
    );

    res.json({
      stats: {
        totalEarnings: total[0].total,
        thisMonth: thisMonth[0].total,
        pending: pending[0].total
      },
      transactions,
      monthlyEarnings
    });
  } catch (err) {
    console.error('Get earnings error:', err);
    res.status(500).json({ error: 'Server error.' });
  }
};

File: backend-spring/src/main/java/com/agrivalue/controller/FarmerController.java
package com.agrivalue.controller;

import com.agrivalue.entity.Order;
import com.agrivalue.entity.Product;
import com.agrivalue.repository.*;
import com.agrivalue.security.JwtUserPrincipal;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/farmer")
public class FarmerController {

    private final ProductRepository productRepo;
    private final OrderItemRepository orderItemRepo;
    private final OrderRepository orderRepo;
    private final UserRepository userRepo;
    private final ProductImageRepository imageRepo;

    public FarmerController(ProductRepository productRepo, OrderItemRepository orderItemRepo,
                            OrderRepository orderRepo, UserRepository userRepo, ProductImageRepository imageRepo) {
        this.productRepo = productRepo;
        this.orderItemRepo = orderItemRepo;
        this.orderRepo = orderRepo;
        this.userRepo = userRepo;
        this.imageRepo = imageRepo;
    }

    @GetMapping("/dashboard")
    public ResponseEntity<?> getDashboardStats(@AuthenticationPrincipal JwtUserPrincipal principal) {
        Integer farmerId = principal.getId();

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalProducts", productRepo.countByFarmerId(farmerId));
        stats.put("activeProducts", productRepo.countByFarmerIdAndStatus(farmerId, Product.ProductStatus.active));
        stats.put("totalOrders", orderItemRepo.countByFarmerId(farmerId));
        stats.put("pendingOrders", orderItemRepo.countByFarmerIdAndStatus(farmerId, Order.OrderStatus.pending));
        stats.put("totalEarnings", orderItemRepo.getTotalEarningsByFarmerId(farmerId));

        List<Map<String, Object>> recentOrders = orderItemRepo.findByFarmerIdOrderByCreatedAtDesc(farmerId)
                .stream().limit(5).map(oi -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("price_at_purchase", oi.getPriceAtPurchase());
                    m.put("quantity", oi.getQuantity());
                    productRepo.findById(oi.getProductId()).ifPresent(p -> m.put("product_name", p.getName()));
                    orderRepo.findById(oi.getOrderId()).ifPresent(o -> {
                        m.put("order_date", o.getCreatedAt());
                        m.put("order_status", o.getStatus().name());
                        userRepo.findById(o.getBuyerId()).ifPresent(u -> m.put("buyer_name", u.getName()));
                    });
                    return m;
                }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("stats", stats, "recentOrders", recentOrders));
    }

    @GetMapping("/inventory")
    public ResponseEntity<?> getInventory(@AuthenticationPrincipal JwtUserPrincipal principal) {
        List<Product> products = productRepo.findByFarmerIdOrderByQuantityAsc(principal.getId());

        int totalItems = products.size();
        int lowStock = (int) products.stream().filter(p -> p.getQuantity().compareTo(BigDecimal.ZERO) > 0 && p.getQuantity().compareTo(new BigDecimal("10")) <= 0).count();
        int outOfStock = (int) products.stream().filter(p -> p.getQuantity().compareTo(BigDecimal.ZERO) <= 0).count();

        List<Map<String, Object>> result = products.stream().map(p -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", p.getId()); m.put("name", p.getName()); m.put("category", p.getCategory().name());
            m.put("quantity", p.getQuantity()); m.put("unit", p.getUnit()); m.put("price", p.getPrice());
            m.put("status", p.getStatus().name());
            imageRepo.findFirstByProductIdAndIsPrimaryTrue(p.getId()).ifPresent(img -> m.put("image", img.getImageUrl()));
            return m;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("stats", Map.of("totalItems", totalItems, "lowStock", lowStock, "outOfStock", outOfStock), "products", result));
    }

    @GetMapping("/earnings")
    public ResponseEntity<?> getEarnings(@AuthenticationPrincipal JwtUserPrincipal principal) {
        Integer farmerId = principal.getId();

        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("totalEarnings", orderItemRepo.getTotalEarningsByFarmerId(farmerId));
        stats.put("thisMonth", orderItemRepo.getTotalEarningsByFarmerId(farmerId)); // simplified
        stats.put("pending", orderItemRepo.getPendingEarningsByFarmerId(farmerId));

        List<Map<String, Object>> transactions = orderItemRepo.findByFarmerIdOrderByCreatedAtDesc(farmerId)
                .stream().limit(10).map(oi -> {
                    Map<String, Object> m = new LinkedHashMap<>();
                    m.put("price_at_purchase", oi.getPriceAtPurchase());
                    m.put("quantity", oi.getQuantity());
                    m.put("amount", oi.getPriceAtPurchase().multiply(oi.getQuantity()));
                    productRepo.findById(oi.getProductId()).ifPresent(p -> m.put("product_name", p.getName()));
                    orderRepo.findById(oi.getOrderId()).ifPresent(o -> {
                        m.put("order_date", o.getCreatedAt());
                        m.put("payment_status", o.getPaymentStatus().name());
                        userRepo.findById(o.getBuyerId()).ifPresent(u -> m.put("buyer_name", u.getName()));
                    });
                    return m;
                }).collect(Collectors.toList());

        return ResponseEntity.ok(Map.of("stats", stats, "transactions", transactions, "monthlyEarnings", List.of()));
    }
}
